### PR TITLE
Improve Dai Echidna Tests

### PR DESCRIPTION
**Pull request by naszam**
- Improve Math to check for overflow
- Filter wad seed

### CODE REVIEW

File: contracts/test/DaiEchidnaTest.sol [hunk 1]
```diff
@@ -24,14 +26,18 @@ contract DaiEchidnaTest {
...
     }
     function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x - y) <= x);
+        z = x - y;
+        assert (z <= x); // check if there is a subtraction overflow
     }
 
     /// @dev Test that supply and balance hold on mint
     function mint(uint256 wad) public {
+        wad = wad % uint256(-1);
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:39]: You can wind up causing overflows this way. I'd define a constant `MAX_SUPPLY` and then set `wad = wad % (MAX_SUPPLY - dai.totalSupply());`. The value of `MAX_SUPPLY` should be some reasonable upper bound on what we expect DAI supply could reach. Maybe `10**15 * WAD`, for example.

File: contracts/test/DaiEchidnaTest.sol [hunk 2]
```diff
@@ -41,6 +47,8 @@ contract DaiEchidnaTest {
 
     /// @dev Test that supply and balance hold on burn
     function burn(uint256 wad) public {
+        wad = wad % uint256(-1);
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:50]: Similar thing here, you might try to `burn` more DAI than the holder has. So I might instead do something like `wad = 1+ (wad % holderBalance)`.

File: contracts/test/DaiEchidnaTest.sol [hunk 3]
```diff
@@ -50,6 +58,8 @@ contract DaiEchidnaTest {
 
     /// @dev Test that supply and balance hold on transfer.
     function transfer(uint256 wad) public {
+        wad = wad % uint256(-1);
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:61]: Similar comment--you don't want to try to transfer more than this address holds.

File: contracts/test/DaiEchidnaTest.sol [hunk 4]
```diff
@@ -59,6 +69,8 @@ contract DaiEchidnaTest {
 
     /// @dev Test that supply and balance hold on transferFrom.
     function transferFrom(uint256 wad) public {
+        wad = wad % uint256(-1);
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:72]: Similar comment--you don't want to try to transfer more than the holder has.

File: contracts/test/DaiEchidnaTest.sol [hunk 5]
```diff
@@ -24,14 +26,18 @@ contract DaiEchidnaTest {
...
     function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x - y) <= x);
+        z = x - y;
+        assert (z <= x); // check if there is a subtraction overflow
     }
 
     /// @dev Test that supply and balance hold on mint
     function mint(uint256 wad) public {
+        wad = wad % uint256(-1);
+        if (wad < WAD) wad = (1 + wad) * WAD;
```
[the remark this answers is not in the document]
Comment (kmbarry1) [reply contracts/test/DaiEchidnaTest.sol:40]: (this comment applies in the tests below, as well)

File: contracts/test/DaiEchidnaTest.sol (hunk 1 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 4 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 3 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 2 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 3 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: same as above for `transfer` test.

File: contracts/test/DaiEchidnaTest.sol (hunk 4 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: same as above for `transferFrom` test.

File: contracts/test/DaiEchidnaTest.sol (hunk 2 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol [hunk 6]
```diff
@@ -42,25 +48,28 @@ contract DaiEchidnaTest {
     /// @dev Test that supply and balance hold on burn
     function burn(uint256 wad) public {
         uint256 supply = dai.totalSupply();
-        uint256 holderBalance = dai.balanceOf(holder);
+        uint256 holderBalance = dai.balanceOf(holder) == 0 ? 1 : dai.balanceOf(holder);
+        wad = 1 + wad % holderBalance;
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:52]: I think this still isn't quite right--if `dai.balanceOf(holder) == 0`, you can't `burn` any amount other than zero. So I'd do something more like:

```
holderBalance = dai.balanceOf(holder);
wad = holderBalance == 0 ? 0 : 1 + wad % holderBalance
```

File: contracts/test/DaiEchidnaTest.sol [hunk 7]
```diff
@@ -42,25 +48,28 @@ contract DaiEchidnaTest {
...
         dai.burn(holder, wad);
         assert(dai.balanceOf(holder) == sub(holderBalance, wad));
         assert(dai.totalSupply() == sub(supply, wad));
     }
 
-    /// @dev Test that supply and balance hold on transfer.
+    /// @dev Test that supply and balance hold on transfer
     function transfer(uint256 wad) public {
-        uint256 thisBalance = dai.balanceOf(address(this));
+        uint256 thisBalance = dai.balanceOf(address(this)) == 0 ? 1 : dai.balanceOf(address(this));
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:60]: similar comment

File: contracts/test/DaiEchidnaTest.sol [hunk 8]
```diff
@@ -42,25 +48,28 @@ contract DaiEchidnaTest {
...
         assert(dai.balanceOf(address(this)) == sub(thisBalance, wad));
         assert(dai.balanceOf(holder) == add(holderBalance, wad));
     }
 
-    /// @dev Test that supply and balance hold on transferFrom.
+    /// @dev Test that supply and balance hold on transferFrom
     function transferFrom(uint256 wad) public {
         uint256 thisBalance = dai.balanceOf(address(this));
-        uint256 holderBalance = dai.balanceOf(holder);
+        uint256 holderBalance = dai.balanceOf(holder) == 0 ? 1 : dai.balanceOf(holder);
```
Comment (kmbarry1) [inline contracts/test/DaiEchidnaTest.sol:71]: similar coment

File: contracts/test/DaiEchidnaTest.sol (hunk 6 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 7 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 8 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 3 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

File: contracts/test/DaiEchidnaTest.sol (hunk 4 above)
Comment (naszam) [reply contracts/test/DaiEchidnaTest.sol]: Done.

**Review by kmbarry1 — APPROVED** [review]
LGTM

**Pull request merged**